rfc/2023/2023-05-01-va-rfc-frontend-performance.md

```diff
@@ -0,0 +1,172 @@
+
+Datadog and [Google Lighthouse](https://developer.chrome.com/docs/lighthouse/) help us understand the Veteran experience with VA.gov by measuring frontend performance. The information provided can help teams identify how to improve website performance for Veterans using various devices and internet connection speeds. These measurements can also help teams define and adopt a budget for how much HTML, CSS and JavaScript should be served per application and per pageview.
+
+### Performance reports show large amounts of unused code, sometimes over 40% of an application, slowing page load times
+
+We can address many challenges to frontend performance on VA.gov today. Every performance report on the [Frontend Performance Dashboard](https://department-of-veterans-affairs.github.io/veteran-facing-services-tools/frontend-support-dashboard/lighthouse-performance-report/) links to a "Treemap", a visualization of what JavaScript libraries and components are included and estimates for how much is unused or duplicated. For example, the treemap for the [Va.gov homepage's report](https://vetsgov-website-builds-s3-upload-test.s3-us-gov-west-1.amazonaws.com/lighthouse/homepage.html) shows that of 3.59MB of JavaScript, 1.55MB, or about **43% is unused**.
+
+### Redundant code is another major source of bloat, slowing down page load times, and is preventable
+
+Further digging into existing performance reports, it appears that most, if not all, pages have three JavaScript bundles, and each bundle has its own copy of React. It appears that **pages have redundant copies of the same code libraries** (e.g., multiple copies of React and ReactDOM). In part, this is due to pages including `web-components.entry.js` (508.752 KiB) and `vendor.entry.js` (349.064 KiB) in addition to an application-specific bundle. The tools we use today can [prevent this duplication](https://webpack.js.org/guides/code-splitting/#prevent-duplication).
+
+A sample of performance scores is available below under [Performance Scores](#performance-scores).
+
+## Design
+
+> Explain the proposed design in enough detail so that a team member will fully understand the implementation. Include a diagram (in the `images` dir) as needed to convey your plans. Use active voice, present tense, and decisive language.
+
+### Set performance expectations and provide tools to monitor performance
+
+The Platform should provide guidance and technical resources to empower VFS teams to adopt performance budgets. While the Platform can provide knowledge and set expectations around performance, VFS teams should take responsibility for it. Providing access to Datadog's RUM will give teams information about performance via Core Web Vitals, which can help them come up with performance budget targets. Teams are already gaining access to Datadog, and adding RUM to measure Core Web Vitals is a reasonable next step.
+
+### Address performance issues that affect all applications
+
```

Comment:
💯 a CDN or back-end redis caching could be a tremendous performance boost to the feature toggles requests

Reply:
We tried to push a CDN a few years ago and it was denied, I think, for security reasons (?). Might be worth trying again though

Reply:
You were probably closer to this, @timwright12, so forgive me if I'm wrong, but I got the impression that:

- security policy forbids storing VA data outside the TIC
- there was an unresolved question about whether this referred to "original" data or any derivatives used in operations/served to end users
- Elijah sent a few pings up the chain
- those requests simply went unanswered.

I don't know if that last part is because of overwork, because the implications are too serious for a brief answer, because miscommunication, etc.

A CDN would be absolutely transformative for a number of teams and for the organization as a whole. Among other things, it would seriously affect the calculus for some of the decisions we (CMS team) would make going forward about architecture.

Reply:
I think a CDN proposal could be another RFC. I tried to keep these RFCs focused on a few areas, though there are certainly more things that can be explored.

To me, the performance issues with the feature toggles endpoint is also worth a discussion. It looks like identity folks were considering an [RFC for feature toggles](https://github.com/department-of-veterans-affairs/va.gov-team/issues/55443). Is there a writeup of the past investigations into feature toggles performance? 


I glanced at the [feature toggle controller](https://github.com/department-of-veterans-affairs/vets-api/blob/master/app/controllers/v0/feature_toggles_controller.rb) and wonder if we've measured many DB/Redis calls occur during that `collect` loop, etc.

Reply:
Added some discussion under `Alternatives`, though I noted that a CDN could be complementary to the proposed approach. I still think a CDN could be its own RFC